# Retail Sales Data Analysis & Customer Segmentation

## Level 3: Feature Engineering

Feature Engineering is the process of creating new variables from existing data to improve analytical capabilities and generate deeper business insights.

### Objective

The objective of this phase is to transform cleaned transactional data into meaningful business-oriented features that enhance analytical capabilities.

These engineered features will support revenue analysis, trend identification, customer behavior analysis, customer segmentation, Customer Lifetime Value (CLV) evaluation, and business decision-making throughout the subsequent stages of the project.

---

### Environment Setup

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Environment Ready")

Environment Ready


All required libraries were successfully imported and the analysis environment is ready.

---

### Loading Cleaned Dataset

In [3]:
df = pd.read_csv(
    "../outputs/cleaned_online_retail.csv",
    parse_dates=["InvoiceDate"]
)

The cleaned dataset is loaded successfully.

---

### Dataset Overview Before Feature Creation

In [4]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.shape

(392692, 8)

The dataset contains 392,692 transaction records and 8 columns.

---

### Revenue Feature Creation

In [6]:
df["Revenue"] = df["Quantity"] * df["UnitPrice"]

A Revenue feature is successfully created by multiplying Quantity and UnitPrice.

This feature represents the monetary value generated by each transaction and will serve as the primary metric for sales performance analysis, customer value evaluation, Pareto analysis, RFM modeling, customer segmentation, and Customer Lifetime Value (CLV) assessment.

---

### Revenue Feature Validation

In [7]:
df[["Quantity","UnitPrice","Revenue"]].head()

,Quantity,UnitPrice,Revenue
0,6,2.55,15.30
1,6,3.39,20.34
2,8,2.75,22.00
3,6,3.39,20.34
4,6,3.39,20.34


The Revenue feature is validated successfully. Sample records confirm that transaction revenue is accurately calculated as the product of Quantity and Unit Price.

The feature is ready for subsequent analytical and business intelligence tasks.

---

### Year Feature Creation

In [8]:
df["Year"] = df["InvoiceDate"].dt.year

A Year feature is successfully extracted from the transaction date.

This feature supports time-based analysis and allows business performance comparisons across different years.

---

### Month Feature Creation

In [9]:
df["Month"] = df["InvoiceDate"].dt.month

A Month feature is created successfully to support monthly sales analysis and seasonality identification.

This feature will help uncover patterns in customer purchasing behavior throughout the year.

---

### Month Name Feature Creation

In [10]:
df["MonthName"] = df["InvoiceDate"].dt.month_name()

The MonthName feature is generated successfully to provide business-friendly month labels.

This improves the readability of reports, visualizations, and business presentations.

---

### Quarter Feature Creation

In [11]:
df["Quarter"] = df["InvoiceDate"].dt.quarter

A Quarter feature is created successfully to support quarterly business performance analysis.

This feature enables revenue and sales comparisons across different quarters of the year.

---

### Day Of Week Feature Creation

In [12]:
df["DayOfWeek"] = df["InvoiceDate"].dt.day_name()

The DayOfWeek feature is generated successfully to identify customer purchasing patterns across weekdays.

This feature will help determine which days contribute most to business activity and revenue generation.

---

### Invoice Month-Year Feature Creation

In [13]:
df["InvoiceMonthYear"] = (
    df["InvoiceDate"]
      .dt.to_period("M")
      .astype(str)
)

The InvoiceMonthYear feature is successfully created by combining month and year information.

This feature simplifies monthly trend analysis and supports chronological reporting of sales performance.

---

### Weekend Indicator Creation

In [14]:
df["IsWeekend"] = (
    df["InvoiceDate"]
      .dt.dayofweek >= 5
)

A weekend indicator feature is successfully created to distinguish between weekday and weekend transactions.

This feature will support analysis of customer purchasing behavior across different periods of the week.

---

### Verifying Newly Created Features

In [15]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue,Year,Month,MonthName,Quarter,DayOfWeek,InvoiceMonthYear,IsWeekend
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom,15.30,2010,12,December,4,Wednesday,2010-12,False
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,December,4,Wednesday,2010-12,False
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom,22.00,2010,12,December,4,Wednesday,2010-12,False
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,December,4,Wednesday,2010-12,False
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom,20.34,2010,12,December,4,Wednesday,2010-12,False


In [16]:
df.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue,Year,Month,MonthName,Quarter,DayOfWeek,InvoiceMonthYear,IsWeekend
196453,561879,47566,PARTY BUNTING,16,2011-07-31 12:43:00,4.95,13722.0,United Kingdom,79.20,2011,7,July,3,Sunday,2011-07,True
94192,548666,22502,PICNIC BASKET WICKER SMALL,3,2011-04-01 14:58:00,5.95,13124.0,United Kingdom,17.85,2011,4,April,2,Friday,2011-04,False
151554,556111,23168,CLASSIC CAFE SUGAR DISPENSER,12,2011-06-09 09:16:00,1.25,13004.0,United Kingdom,15.00,2011,6,June,2,Thursday,2011-06,False
220215,564476,21671,RED SPOT CERAMIC DRAWER KNOB,12,2011-08-25 12:57:00,1.25,12610.0,Italy,15.00,2011,8,August,3,Thursday,2011-08,False
143580,555124,85049G,CHOCOLATE BOX RIBBONS,2,2011-05-31 15:26:00,1.25,18204.0,United Kingdom,2.50,2011,5,May,2,Tuesday,2011-05,False
142533,554990,15060B,FAIRY CAKE DESIGN UMBRELLA,8,2011-05-29 13:57:00,3.75,16567.0,United Kingdom,30.00,2011,5,May,2,Sunday,2011-05,True
141206,554835,37370,RETRO COFFEE MUGS ASSORTED,12,2011-05-26 16:11:00,1.25,15241.0,United Kingdom,15.00,2011,5,May,2,Thursday,2011-05,False
65145,545060,21164,HOME SWEET HOME METAL SIGN,1,2011-02-27 13:47:00,2.95,15547.0,United Kingdom,2.95,2011,2,February,1,Sunday,2011-02,True
242543,567095,82482,WOODEN PICTURE FRAME WHITE FINISH,6,2011-09-16 13:03:00,2.95,17049.0,United Kingdom,17.70,2011,9,September,3,Friday,2011-09,False
10200,537624,21827,EIGHT PIECE CREEPY CRAWLIE SET,4,2010-12-07 14:41:00,1.25,12748.0,United Kingdom,5.00,2010,12,December,4,Tuesday,2010-12,False


In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 392692 entries, 0 to 392691
Data columns (total 16 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   InvoiceNo         392692 non-null  int64         
 1   StockCode         392692 non-null  str           
 2   Description       392692 non-null  str           
 3   Quantity          392692 non-null  int64         
 4   InvoiceDate       392692 non-null  datetime64[us]
 5   UnitPrice         392692 non-null  float64       
 6   CustomerID        392692 non-null  float64       
 7   Country           392692 non-null  str           
 8   Revenue           392692 non-null  float64       
 9   Year              392692 non-null  int32         
 10  Month             392692 non-null  int32         
 11  MonthName         392692 non-null  str           
 12  Quarter           392692 non-null  int32         
 13  DayOfWeek         392692 non-null  str           
 14  InvoiceMonthYea

In [18]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country', 'Revenue', 'Year', 'Month',
       'MonthName', 'Quarter', 'DayOfWeek', 'InvoiceMonthYear', 'IsWeekend'],
      dtype='str')

All engineered features are successfully created and integrated into the dataset.

Dataset validation confirms that the new variables are available for analysis and can be utilized in Exploratory Data Analysis (EDA), Pareto Analysis, RFM Analysis, Customer Segmentation, and Customer Lifetime Value (CLV) evaluation.

---

### Exporing Feature Engineered Dataset

In [19]:
df.to_csv(
    "../outputs/feature_engineered_retail.csv",
    index=False
)

print("Feature-engineered dataset saved successfully.")

Feature-engineered dataset saved successfully.


The feature-engineered dataset is successfully exported and stored for future analytical stages.

The dataset now contains both original transactional attributes and newly engineered business features, making it fully prepared for exploratory analysis and advanced customer analytics.

---

### Summary

#### Feature Engineering Activities

The following business-oriented features were created from the cleaned transactional dataset:

- Revenue (Quantity × Unit Price)
- Year
- Month
- Month Name
- Quarter
- Day of Week
- Invoice Month-Year
- Weekend Indicator (IsWeekend)

These features were derived from transaction and date-related information to enhance analytical capabilities and support advanced customer analytics.

---

#### Feature Enhancement Impact

The engineered features transformed raw transactional records into a richer analytical dataset by introducing:

- Revenue-based business metrics
- Time-series analysis capabilities
- Seasonal and quarterly analysis support
- Customer purchasing pattern analysis
- Transaction timing insights

These enhancements enable deeper exploration of sales trends and customer behavior.

---

#### Dataset Impact

| Metric | Value |
|----------|----------|
| Records | 392,692 |
| Original Columns | 8 |
| Engineered Features Added | 8 |
| Final Columns | 16 |

The dataset now contains both transactional attributes and business-oriented analytical features.

---

#### Business Implications

The engineered features provide the foundation for:

- Revenue Trend Analysis
- Product Performance Analysis
- Customer Purchase Behavior Analysis
- Pareto Analysis
- RFM Analysis
- Customer Lifetime Value (CLV) Analysis
- Customer Segmentation

These features improve the dataset's ability to support customer intelligence and business decision-making.

---

#### Conclusion

The Feature Engineering phase successfully enriched the cleaned retail dataset with meaningful business-oriented features. By transforming raw transactional information into analytical variables, the dataset is now better equipped for uncovering sales trends, customer behavior patterns, and revenue-driving insights.

---

#### Next Step

**Level 4: Exploratory Data Analysis (EDA)**

The next phase will focus on analyzing sales performance, customer purchasing behavior, product contribution, revenue concentration, and geographic trends to generate actionable business insights.